beacon-chain/rpc/apimiddleware/endpoints.go

```diff
@@ -0,0 +1,316 @@
+		},
+		{
+			Url:         "/eth/v1/node/peer_count",
+			GetResponse: &peerCountResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+		{
+			Url:         "/eth/v1/node/version",
+			GetResponse: &versionResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+		{
+			Url:         "/eth/v1/node/syncing",
+			GetResponse: &syncingResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+		{
+			Url: "/eth/v1/node/health",
+			Err: &gateway.DefaultErrorJson{},
+		},
+		{
+			Url:         "/eth/v1/debug/beacon/states/{state_id}",
+			GetResponse: &beaconStateResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+			Hooks: gateway.HookCollection{
+				CustomHandlers: []gateway.CustomHandler{handleGetBeaconStateSsz},
+			},
+		},
+		{
+			Url:         "/eth/v1/debug/beacon/heads",
+			GetResponse: &forkChoiceHeadsResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+		{
+			Url:         "/eth/v1/config/fork_schedule",
+			GetResponse: &forkScheduleResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+		{
+			Url:         "/eth/v1/config/deposit_contract",
+			GetResponse: &depositContractResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+		{
+			Url:         "/eth/v1/config/spec",
+			GetResponse: &specResponseJson{},
+			Err:         &gateway.DefaultErrorJson{},
+		},
+	}
+}
+
+// https://ethereum.github.io/eth2.0-APIs/#/Beacon/submitPoolAttestations expects posting a top-level array.
+// We make it more proto-friendly by wrapping it in a struct with a 'data' field.
+func wrapAttestationsArray(endpoint gateway.Endpoint, _ http.ResponseWriter, req *http.Request) gateway.ErrorJson {
+	if _, ok := endpoint.PostRequest.(*submitAttestationRequestJson); ok {
+		atts := make([]*attestationJson, 0)
+		if err := json.NewDecoder(req.Body).Decode(&atts); err != nil {
+			e := fmt.Errorf("could not decode attestations array: %w", err)
+			return &gateway.DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}
+		}
+		j := &submitAttestationRequestJson{Data: atts}
+		b, err := json.Marshal(j)
+		if err != nil {
+			e := fmt.Errorf("could not marshal wrapped attestations array: %w", err)
+			return &gateway.DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}
+		}
+		req.Body = ioutil.NopCloser(bytes.NewReader(b))
+	}
+	return nil
+}
+
+// Posted graffiti needs to have length of 32 bytes, but client is allowed to send data of any length.
+func prepareGraffiti(endpoint gateway.Endpoint, _ http.ResponseWriter, _ *http.Request) gateway.ErrorJson {
+	if block, ok := endpoint.PostRequest.(*beaconBlockContainerJson); ok {
+		b := bytesutil.ToBytes32([]byte(block.Message.Body.Graffiti))
+		block.Message.Body.Graffiti = hexutil.Encode(b[:])
+	}
+	return nil
+}
+
+func handleGetBeaconStateSsz(m *gateway.ApiProxyMiddleware, endpoint gateway.Endpoint, writer http.ResponseWriter, request *http.Request) (handled bool) {
```

Comment:
The implementation of fetching the SSZ version of the beacon states requires changes to very specific parts of the request. Using callbacks here would make the whole callback functionality too granular. Therefore I decided to allow providing custom handlers that essentially replace the default request handling logic. This should be a last resort when implementing an endpoint. 

Because some parts of the default logic are still the same, I decided to export some functions from the shared gateway package so that they can be reused.

An analogous endpoint is coming in https://github.com/prysmaticlabs/prysm/pull/8970.